### PR TITLE
Sine distribution

### PR DESCRIPTION
**Pull request by williamjameshandley**
Added a new 'sine' distribution, which has the pdf: 1/2 * sin(x) for x in [0,pi] . This is useful for example for a distribution on a polar angle giving samples uniformly distributed on the sphere

### CODE REVIEW

**Comment by chrisb83** [comment]
yes, the rst file needs to be added manually.

Where is this distribution used?

Regarding the implementation: 
- cdf ist wrong, `sin(x/2)**2` ist not a primitive of `sin(x)`, it should be `1 - cos(x/2)`?
- `_rvs` needs to be added to generate random variates (efficiently)

**Comment by williamjameshandley** [comment]
* added `_rvs`
* corrected `_cdf`
* Added rst documentation

> Where is this distribution used?

I've used it in my research when generating points randomly on a sphere in spherical polar coordinates:

e.g:
```
size=100
phi = scipy.stats.uniform(0,2*np.pi).rvs(size)
theta = scipy.stats.sine().rvs(size)
```

**Comment by chrisb83** [comment]
The distribution is not easy to find in references, I found an article (http://ww2.amstat.org/sections/srms/proceedings/papers/1987_060.pdf) that mentions it. 

Before I review the code in more detail, let´s also wait for feedbck from others if this distribution should be included.

**Comment by chrisb83** [comment]
@josef-pkt do you have an opinion whether this distribution should be added to scipy?

**Comment by josef-pkt** [comment]
In general I don't see much reason why we should restrict adding new distributions, i.e. there needs to be a good reason why NOT to add it.

I don't know the sine distribution either. The main question would be whether this is supposed to be a circular distribution and we run into the same problem as for vonmises_line versus vonmises versus having a real circular distribution.

**Comment by josef-pkt** [comment]
If it is for a fixed interval like [0, pi], then I don't see a problem. However, loc and scale would shift and scale the interval, and there is no parameter to change the shape withing the [0, pi] interval (i.e. it's still similar to vonmises_line).

My guess would be, that to make it useful, the distribution would need to change shape in the given [0, pi] interval.

**Comment by josef-pkt** [comment]
Or phrased differently:
How do you use this when the shape is not the default, i.e. either shifted or with more or less dispersion?

**Comment by williamjameshandley** [comment]
I had modelled this on the `cosine` distribution, and had assumed that the shape would adjust accordingly if you provided alternative shape parameters (shift and dispersion). Am I incorrect? The correct behaviour would be for the interval to change shape with shift and dispersion, so that the `sine` distribution always has zeros at the bounds of its support. If the current version does not implement this, is there an example of a distribution that has similar behaviour?

**Comment by josef-pkt** [comment]
loc and scale are generic transformation of the data or distribution. The bounds are correspondingly shifted.
e.g. 
scale=2 means that the support is [0, 2 * pi] and the shape of the pdf is correspondingly stretched to that interval.
loc=1 just would shift the interval, i.e. support is [1, 1 + pi] with the same shape of the distribution as with loc=0.

Because the entire distribution including the support is shifted and stretched with changes in loc and scale, we cannot change the shape of the distribution for the fixed interval [0, pi].

**Comment by williamjameshandley** [comment]
OK. In which case, the current behaviour is that which is desired. One would not want the sine distribution to change its shape with respect to the domain. If it narrowed, then the distribution would become negative, and it's difficult to imagine a use-case where it were wider.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -5959,6 +5959,48 @@ def _entropy(self):
...
+    %(example)s
+
+    """
+    def _pdf(self, x):
+        # sin.pdf(x) = 1/2 * sin(x)
+        return 1.0/2*(np.sin(x))
+
+    def _rvs(self):
+        z = self._random_state.uniform(size=self._size)
+        return 2*np.arcsin(np.sqrt(z))
```
Comment (chrisb83) [inline scipy/stats/_continuous_distns.py]: I don't understand where the sqrt term comes from. Simply inverting the cdf gives arccos(...).

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -5959,6 +5959,48 @@ def _entropy(self):
...
+
+    %(example)s
+
+    """
+    def _pdf(self, x):
+        # sin.pdf(x) = 1/2 * sin(x)
+        return 1.0/2*(np.sin(x))
+
+    def _rvs(self):
+        z = self._random_state.uniform(size=self._size)
```
Comment (chrisb83) [inline scipy/stats/_continuous_distns.py]: Can you add _ppf method as well? (That could actually be used to generate rvs using the default method.)

File: doc/source/tutorial/stats/continuous_sine.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+Sine Distribution
+===================
+
+Polar-angle distribution required for ensuring uniform distribution over a sphere.
+
+.. math::
+   :nowrap:
+
+    \begin{eqnarray*} f\left(x\right) & = & \frac{1}{2}\sin \left(x\right) I_{\left[0,\pi\right]}\left(x\right)\\ F\left(x\right) & = & \frac{1}{2}\left[1-\cos(x)\right]I_{\left[0,\pi\right]}\left(x\right)+I_{\left(\pi,\infty\right)}\left(x\right)\\ G\left(\alpha\right) & = & 2\sin^{-1}\left(\sqrt{x}\right)\\ M\left(t\right) & = & \frac{1+e^{\pi t}}{2\left(1+t^2\right)}\\ \mu=m_{d}=m_{n} & = & \frac{\pi}{2}\\ \mu_{2} & = & \frac{\pi^2}{4}-2\\ \gamma_{1} & = & 0\\ \gamma_{2} & = & -2\frac{\pi^4-96}{\left(\pi^2-8\right)^2}\end{eqnarray*}
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_sine.rst]: Can this be written in a more readable way :)

File: scipy/stats/tests/test_continuous_basic.py [hunk 4]
```diff
@@ -50,7 +50,8 @@
 
 distslow = ['rdist', 'gausshyper', 'recipinvgauss', 'ksone', 'genexpon',
             'vonmises', 'vonmises_line', 'mielke', 'semicircular',
-            'cosine', 'invweibull', 'powerlognorm', 'johnsonsu', 'kstwobign']
+            'cosine', 'sine', 'invweibull', 'powerlognorm', 'johnsonsu',
```
Comment (chrisb83) [inline scipy/stats/tests/test_continuous_basic.py]: Why did you add it Here?

File: doc/source/tutorial/stats/continuous_sine.rst [hunk 5]
```diff
@@ -0,0 +1,19 @@
+
+.. _continuous-sine:
+
+Sine Distribution
+===================
+
+Polar-angle distribution required for ensuring uniform distribution over a sphere.
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_sine.rst]: Can you add a reference? Not a must but nice to have

File: doc/source/tutorial/stats/continuous_sine.rst (hunk 3 above)
Comment (ilayn) [reply doc/source/tutorial/stats/continuous_sine.rst]: Indeed 😃 

```latex

\begin{align*}
f(x) & = \frac{1}{2}\sin{(x)} I_{[0,\pi]}(x)\\
F(x) & = \frac{1}{2}[1-\cos(x)]I_{[0,\pi]}(x)+I_{(\pi,\infty)}(x)\\
G(\alpha) & = 2\sin^{-1}(\sqrt{x})\\
M(t) & = \frac{1+e^{\pi t}}{2(1+t^2)}\\
\mu=m_{d}=m_{n} & = \frac{\pi}{2}\\
\mu_{2} & = \frac{\pi^2}{4}-2\\
\gamma_{1} & = 0\\
\gamma_{2} & = -2\frac{\pi^4-96}{(\pi^2-8)^2}
\end{align*}
```
If possible please avoid using `eqnarray` completely.

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -5959,6 +5959,48 @@ def _entropy(self):
...
+
+    for :math:`0 \le x \le \pi`.
+
+    %(after_notes)s
+
+    %(example)s
+
+    """
+    def _pdf(self, x):
+        # sin.pdf(x) = 1/2 * sin(x)
```
[the remark this answers is not in the document]
Comment (williamjameshandley) [reply scipy/stats/_continuous_distns.py]: Done

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -5959,6 +5959,48 @@ def _entropy(self):
...
+        return 1.0/2*(np.sin(x))
```
[the remark this answers is not in the document]
Comment (williamjameshandley) [reply scipy/stats/_continuous_distns.py]: Done

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (williamjameshandley) [reply scipy/stats/_continuous_distns.py]: `2*arcsin(sqrt(x)) = arccos(1-2*x)`. Re-written as the latter

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (williamjameshandley) [reply scipy/stats/_continuous_distns.py]: Added _ppf, and removed _rvs

File: doc/source/tutorial/stats/continuous_sine.rst (hunk 3 above)
Comment (williamjameshandley) [reply doc/source/tutorial/stats/continuous_sine.rst]: Done. I was merely following the format in cosine. I thought it seemed a little odd to have that as the convention.

File: scipy/stats/tests/test_continuous_basic.py (hunk 4 above)
Comment (williamjameshandley) [reply scipy/stats/tests/test_continuous_basic.py]: Removed

File: doc/source/tutorial/stats/continuous_sine.rst (hunk 5 above)
Comment (williamjameshandley) [reply doc/source/tutorial/stats/continuous_sine.rst]: There isn't a particularly good one that I can think of, and none of the others in `doc/source/tutorial/stats/` seem to have a reference.

Comment (chrisb83) [reply doc/source/tutorial/stats/continuous_sine.rst]: OK, i wanted to add this comment to the doc string of the distribution. Can you add the refernece i posted in the thread. It states the moments you use in _stats, so that ist useful to add.

File: scipy/stats/_continuous_distns.py [hunk 8]
```diff
@@ -5959,6 +5959,46 @@ def _entropy(self):
...
+    %(before_notes)s
+
+    Notes
+    -----
+    The sine distribution is useful for polar angles.
+    The probability density function for `sine` is:
+
+    .. math::
+
+        f(x) = \frac{1}{2} \sin(x)
```
Comment (chrisb83) [inline scipy/stats/_continuous_distns.py]: Probably simpler to write 0.5 instead of frac, there was a discussion to minimize LaTeX expressions some time ago

**Comment by chrisb83** [comment]
Thanks for pushing the changes. PR looks good in my view, I added two minor comment. 

Not sure why the 2.7 build fails.

**Comment by williamjameshandley** [comment]
I updated the documentation as requested, and chose to put the reference into `scipy/stats/_continuous_distns.py` rather than `doc/source/tutorial/stats/continuous_sine.rst`, as this seems more consistent with the rest of the referencing.

I also have no idea why the build fails for 2.7. I spent some time looking at it, and it's not obviously related to the changes I've made.

**Comment by ilayn** [comment]
Nevermind the unrelated failures they can come also through other PRs or NumPy changes but I think you are missing the tests for sine or forgot to add.

**Comment by ilayn** [comment]
Is there anything else to be done for this for the merge?

**Comment by josef-pkt** [comment]
Looks good to me, but other methods should be added if there is a closed form expression, e.g. `sf` and `isf`, and maybe `moments` if those are available for all moments.
Given that ppf is an analytical expression, there is most likely not much to gain with an explicit rvs.

**Comment by josef-pkt** [comment]
@chrisb83  We still need sf and isf if available (in general)
1-cdf  can be only of the order of floating point at 1, i.e. something like 1 - (1-1e-16)
If sf is defined, then we can get values far out in the tail with small probability.

```
>>> stats.norm.sf(15)
3.6709661993126986e-51
>>> 1 - stats.norm.cdf(15)
0.0
```

This might not be relevant in this case, but I think it's still better to add it to avoid having to think about the possibility of precision loss. (I haven't checked for this case.)

**Comment by chrisb83** [comment]
@josef-pkt agree, if you can rewrite `sf` in a way that is better from a numerical point of view than `1 - cdf`. I don't know if this can be achieved here (not obvious in my view).

If this is not the case, I would say they don't need to be added (this seems to be the current guidance on https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html: "Rarely would you override _isf, _sf or _logsf, but you could.")

**Comment by josef-pkt** [comment]
> "Rarely would you override _isf, _sf or _logsf, but you could."

I think that's misleading and it would be better to remove it.
Over the years missing methods have been added (when available) to improve precision in many cases.

sf and isf look easy to add in this case.

**Comment by WarrenWeckesser** [comment]
Use the [half angle formulas](https://en.wikipedia.org/wiki/List_of_trigonometric_identities#Half-angle_formulae) to write

    CDF(x) = sin(x/2)**2
    SF(x) = cos(x/2)**2

Then the CDF can be computed more accurately near x=2, and the SF can be computed more accurately near π.

**Comment by chrisb83** [comment]
@williamjameshandley thanks for continuing to work on this, I think we are almost there, can you add `isf` as well?

And remove the link in the reference.

**Comment by WarrenWeckesser** [comment]
If you invert the formulas that use the half angle indentities, you can improve the `_ppf` and `_isf` methods.  (The subtraction `1 - 2*q` causes severe loss of precision when `q` is very small.)  For example,

```
In [104]: x = 1e-15

In [105]: np.arccos(1 - 2*x)       # PPF as currently implemented
Out[105]: 6.322027276634106e-08

In [106]: 2*np.arcsin(np.sqrt(x))  # PPF computed by inverting the half angle formula for CDF
Out[106]: 6.32455532033676e-08

In [107]: import mpmath

In [108]: mpmath.mp.dps = 200

In [109]: float(mpmath.acos(1 - 2*mpmath.mpf(x)))   # Very high precision calculation
Out[109]: 6.32455532033676e-08
```
When the argument is even small, all precision is lost:
```
In [110]: x = 1e-18

In [111]: np.arccos(1 - 2*x)
Out[111]: 0.0

In [112]: 2*np.arcsin(np.sqrt(x))
Out[112]: 2e-09

In [113]: float(mpmath.acos(1 - 2*mpmath.mpf(x)))
Out[113]: 2e-09
```

**Comment by williamjameshandley** [comment]
OK that's done, although the build is still failing for osx for an apparently unrelated reason.

**Comment by chrisb83** [comment]
OK, looks good to me.

**Comment by WarrenWeckesser** [comment]
@williamjameshandley, sorry I didn't notice this sooner, but this `sine` distribution is a shifted and scaled version of the [`anglit` distribution](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.anglit.html), which is already in SciPy.  So we shouldn't merge this.

I don't know why `anglit` has such a cryptic name; I only noticed this now because I was browsing through the existing distributions looking for something else.

**Comment by williamjameshandley** [comment]
No worries, at least I learned something about how to implement new distributions if ever I need to in the future.

**Comment by ilayn** [comment]
Probably it is from the family of logit, normit, anglit and so on. 

Do we need an alias for this?

**Comment by chrisb83** [comment]
The `anglit` distribution does not have `_isf` and `_sf` methods, so this could be sth to add based on the discussion above.

**Comment by williamjameshandley** [comment]
Should I create a new pull request for that?

**Comment by chrisb83** [comment]
@williamjameshandley yes, if you want to work on that, it is best to open a new PR and close this one.
